test/logflare_web/controllers/api/source_controller_test.exs

```diff
@@ -42,4 +45,107 @@ defmodule LogflareWeb.Api.SourceControllerTest do
       assert_lists_equal(response, expected)
     end
   end
+
+  describe "create/2" do
+    test "creates a new source for an authenticated user", %{
+      conn: conn,
+      users: [u1, _u2]
+    } do
+      name = TestUtils.random_string()
+
+      conn
+      |> login_user(u1)
+      |> post("/api/sources", %{name: name})
+      |> response(204)
+
+      assert Repo.one(
+               from(s in Source,
+                 where: s.user_id == ^u1.id,
+                 where: s.name == ^name,
+                 select: s.name
+               )
+             )
+    end
+  end
+
+  describe "update/2" do
+    test "updates an existing source from a user", %{
+      conn: conn,
+      users: [u1, _u2],
+      sources: [s1 | _]
+    } do
+      name = TestUtils.random_string()
+
+      conn
+      |> login_user(u1)
+      |> patch("/api/sources/#{s1.id}", %{name: name})
+      |> response(204)
+
+      assert name ==
+               Repo.one(
+                 from(s in Source,
+                   where: s.id == ^s1.id,
+                   select: s.name
+                 )
+               )
+    end
+
+    test "returns not found if doesn't own the source", %{
+      conn: conn,
+      users: [u1, _u2],
+      sources: [_s1, _s2, s3]
+    } do
+      conn
+      |> login_user(u1)
+      |> patch("/api/sources/#{s3.id}", %{name: TestUtils.random_string()})
+      |> response(404)
+    end
+  end
+
+  describe "delete/2" do
+    test "deletes an existing source from a user", %{
+      conn: conn,
+      users: [u1, _u2],
+      sources: [s1 | _]
+    } do
+      name = TestUtils.random_string()
+
+      conn
+      |> login_user(u1)
+      |> delete("/api/sources/#{s1.id}", %{name: name})
+      |> response(204)
+
+      assert nil ==
+               Repo.one(
+                 from(s in Source,
+                   where: s.id == ^s1.id,
+                   select: s.name
+                 )
+               )
+    end
+
+    test "returns not found if doesn't own the source", %{
+      conn: conn,
+      users: [u1, _u2],
+      sources: [_s1, _s2, s3]
+    } do
+      conn
+      |> login_user(u1)
+      |> delete("/api/sources/#{s3.id}", %{name: TestUtils.random_string()})
+      |> response(404)
+    end
+  end
+
+  test "changeset errors handled gracefully", %{
```

Comment:
we can merge this test together with update/2 and create/2 blocks, just adds 1 extra line of assertion for each test case.

```
# in setup
conn = login_user(user)


assert %{"error"=> %{"name"=> [ "can't" <>_]}} = conn 
        |> post("/api/sources" , @invalid) 
        |> json_response(422)


assert %{"error"=> %{"name"=> [ "can't" <>_]}} = conn 
        |> put("/api/sources/${s.id}" , @invalid) 
        |> json_response(422)

Reply:
will change 👍